File: objective-c/Database/iOS/library/Classes/AddController.m
// **********************************************************************
//
// Copyright (c) 2003-present ZeroC, Inc. All rights reserved.
//
// **********************************************************************

#import <AddController.h>
#import <Library.h>

@interface AddController ()

@property (nonatomic) id<DemoLibraryPrx> library;

@end

@implementation AddController

@synthesize library;

-(void)viewDidLoad
{
    self.title = @"New Book";
    tableView.allowsSelectionDuringEditing = YES;
}

-(void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    self.navigationItem.leftBarButtonItem.enabled = YES;
    self.navigationItem.rightBarButtonItem.enabled = (book.isbn && book.isbn.length > 0);
    [self setEditing:YES animated:NO];
    [self.navigationItem setHidesBackButton:NO animated:animated];
}

-(void)didReceiveMemoryWarning
{
  [super didReceiveMemoryWarning]; // Releases the view if it doesn't have a superview
  // Release anything that's not essential, such as cached data
}

-(IBAction)save:(id)sender
{
    [self saving:YES];
    [library begin_createBook:book.isbn
                        title:book.title
                      authors:book.authors
                     response:^(id<DemoBookPrx> prx) {
                         [self saving:NO];
             [self.navigationController popViewControllerAnimated:YES];
           }
                    exception:^(ICEException* ex) {
            if([ex isKindOfClass:[DemoBookExistsException class]])
            {
                            [self saving:NO];
              // open an alert with just an OK button
              UIAlertController* alert = [UIAlertController alertControllerWithTitle:@"Error"
                                                                             message:@"That ISBN number already exists"
                                                                      preferredStyle:UIAlertControllerStyleAlert];
              [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil]];
              [self presentViewController:alert animated:YES completion:nil];
              return;
            }

            [super exception:ex];
          }];
}

-(void)startEdit:(DemoBookDescription*)b library:(id<DemoLibraryPrx>)l
{
    self.book = b;
    self.library = l;
    [self setEditing:YES animated:NO];
}

@end

File: swift/iOS/library/AddController.swift
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

import UIKit

class AddController: DetailController {
    var library: LibraryPrx!

    override func viewDidLoad() {
        title = "New Book"
        tableView.allowsSelectionDuringEditing = true
        setEditing(true, animated: false)
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        navigationItem.leftBarButtonItem?.isEnabled = true
        navigationItem.rightBarButtonItem?.isEnabled = !book.isbn.isEmpty
        setEditing(true, animated: false)
        navigationItem.setHidesBackButton(false, animated: animated)
    }

    @IBAction func save(sender _: Any?) {
        saving(true)

        library.createBookAsync(isbn: book.isbn, title: book.title, authors: book.authors).done { _ in
            self.saving(false)
            self.navigationController?.popViewController(animated: true)
        }.catch { error in
            if error is BookExistsException {
                self.saving(false)
                let alert = UIAlertController(title: "Error", message: "That ISBN number already exists",
                                              preferredStyle: .alert)
                alert.addAction(UIAlertAction(title: "OK", style: .default, handler: nil))
                self.present(alert, animated: true, completion: nil)
            } else {
                self.exception(error)
            }
        }
    }

    func startEdit(book: BookDescription, library: LibraryPrx?) {
        self.book = book
        self.library = library
        setEditing(true, animated: false)
    }
}
